### ISSUE TITLE
Sometimes KOReader doesn't wake up after opening sleepcover

### ISSUE THREAD

**Issue by allati** [koreader/koreader#5098]
**KOReader version**: v2019.06
**Device**: Kobo Forma
Sleepcover from Kobo. Not sure if it makes the difference, but I counted 6 magnets in it (https://imgur.com/a/G092yBM). I guess that two magnets on the right keeps cover closed or in it's "origami" position. One magnet in the triangle triggers open/close event. And there are three more magnets doing have no idea what, maybe helping the other two to keep cover closed.

**Steps to reproduce**
Sometimes when I open sleepcover, KOReader is not waking up. Closing and then opening sleepcover again usually wakes it.

In similar-looking issue (#5087) NiLuJe suggested that it may be caused by sleepcover magnets triggering sensors when folded to the back of the device. In my case it is unlikely, because I keep it open at 270° max, it shouldn't be hitting sensors at these angles.

Here are excerpts from crash.log for two recent similar cases:

```
====== Case 1

06/20/19-00:58:26 INFO  Kobo suspend: synced FS
06/20/19-00:58:26 INFO  Kobo suspend: asking for a suspend to RAM . . .
06/20/19-00:58:26 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true

//sleepcover opened, but KOReader keeps sleeping
06/20/19-11:03:14 INFO  Kobo suspend: woke up!
06/20/19-11:03:29 ERROR Kobo suspend: putting device back to sleep, unexpected wakeups: 1
06/20/19-11:03:29 INFO  Kobo suspend: going to sleep . . .
06/20/19-11:03:29 INFO  Kobo suspend: asked the kernel to put subsystems to sleep, ret: true
06/20/19-11:03:31 INFO  Kobo suspend: waited for 2s because of reasons...
06/20/19-11:03:31 INFO  Kobo suspend: synced FS
06/20/19-11:03:31 INFO  Kobo suspend: asking for a suspend to RAM . . .
06/20/19-11:03:31 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true

//sleepcover reopened, KOReader wakes up
06/20/19-11:04:04 INFO  Kobo suspend: woke up!
06/20/19-11:04:11 INFO  Kobo resume: clean up after wakeup

====== Case 2

06/20/19-20:39:55 INFO  Kobo suspend: asking for a suspend to RAM . . .
06/20/19-20:39:55 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true

//sleepcover opened, but KOReader keeps sleeping
06/21/19-16:02:59 INFO  Kobo suspend: woke up!
06/21/19-16:03:14 ERROR Kobo suspend: putting device back to sleep, unexpected wakeups: 1
06/21/19-16:03:14 INFO  Kobo suspend: going to sleep . . .
06/21/19-16:03:14 INFO  Kobo suspend: asked the kernel to put subsystems to sleep, ret: true
06/21/19-16:03:16 INFO  Kobo suspend: waited for 2s because of reasons...
06/21/19-16:03:16 INFO  Kobo suspend: synced FS
06/21/19-16:03:16 INFO  Kobo suspend: asking for a suspend to RAM . . .
06/21/19-16:03:16 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true

//sleepcover reopened, KOReader wakes up
06/21/19-16:03:39 INFO  Kobo suspend: woke up!
06/21/19-16:03:39 INFO  Kobo resume: clean up after wakeup
```

There is one no-so-similar case and I'm not certain about my actions relative to the log time

```
06/25/19-14:20:43 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true
06/25/19-14:20:49 INFO  Kobo suspend: woke up!
06/25/19-14:20:51 INFO  Kobo resume: clean up after wakeup
06/25/19-14:20:51 INFO  Kobo resume: unflagged kernel subsystems for resume, ret: true

// sleepcover closed
06/25/19-14:28:13 INFO  Kobo suspend: going to sleep . . .
06/25/19-14:28:13 INFO  Kobo suspend: asked the kernel to put subsystems to sleep, ret: true
06/25/19-14:28:15 INFO  Kobo suspend: waited for 2s because of reasons...
06/25/19-14:28:15 INFO  Kobo suspend: synced FS
06/25/19-14:28:15 INFO  Kobo suspend: asking for a suspend to RAM . . .
06/25/19-14:28:15 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true

// sleepcover opened, but KOReader keeps sleeping
06/25/19-14:28:22 INFO  Kobo suspend: woke up!
06/25/19-14:28:25 INFO  Kobo resume: clean up after wakeup
06/25/19-14:28:25 INFO  Kobo resume: unflagged kernel subsystems for resume, ret: true

// sleepcover closed and opened again, now wakes up
06/25/19-14:32:08 INFO  Kobo suspend: going to sleep . . .
06/25/19-14:32:08 INFO  Kobo suspend: asked the kernel to put subsystems to sleep, ret: true
06/25/19-14:32:10 INFO  Kobo suspend: waited for 2s because of reasons...
06/25/19-14:32:10 INFO  Kobo suspend: synced FS
06/25/19-14:32:10 INFO  Kobo suspend: asking for a suspend to RAM . . .
06/25/19-14:32:10 INFO  Kobo suspend: ZzZ ZzZ ZzZ? Write syscall returned:  true
06/25/19-14:32:11 INFO  Kobo suspend: woke up!

// closed sleepcover again for the rest of the day
06/25/19-14:32:26 ERROR Kobo suspend: putting device back to sleep, unexpected wakeups: 1
06/25/19-14:32:26 INFO  Kobo suspend: going to sleep . . .
06/25/19-14:32:26 INFO  Kobo suspend: asked the kernel to put subsystems to sleep, ret: true
```

Maybe log level can be increased somehow to provide more data for diagnostics?

Thanks.

**Comment by NiLuJe** [comment]
You can enable verbose debug logging the Developper settings menu of the gear tab (last entry in that menu, from the FM only).

It'll be *very* verbose, but that might be helpful, because the Forma sends key repeat events for the SleepCover -_-". Which we ignore, but, still :D.

Sidebar: that's a lot of magnets :D. IIRC, the hall sensor is somewhere around the top left of the screen, so that'd be the second circle from the left when not doing any origami folds ;p.

**Comment by allati** [comment]
It seems that at least my last case is caused by sleepcover staying closed for about 20 seconds more or less. I saved crash.log at pastebin.com with 2 week expiration: https://pastebin.com/Y54DfHtq

The actions were:
1. launch KOReader
2. Wait several seconds and close sleepcover (~02:41:30)
3. Wait ~20 seconds and open sleepcover (KOReader keeps sleeping)  (~02:41:50)
4. Wait ~10 seconds and close-open sleepcover (KOReader wakes)  (~02:42:00)

**Comment by NiLuJe** [comment]
Huh. It's not actually going to sleep immediately, which means you end up opening the sleepcover and triggering a wakeup in the middle of what ought to already be suspended but isn't => kablooey.

I'll have to dig up some magnets tomorrow. This is highly annoying to test, because I can't do it over USBNet, as it prevents suspend :/.

**Comment by NiLuJe** [comment]
What is KOReader setup to show as a screensaver on your device?

**Comment by allati** [comment]
I use reading progress as screensaver, with all options in the setting menu unchecked, with disabled "delay when exit from screensaver". Screensaver folder, image and message options should have default value as I haven't touched them.

**Comment by NiLuJe** [comment]
Okay, the delayed suspend is perfectly normal, I'd just forgotten about that ;).

On the other hand, I can't reproduce any of this. Playing with a magnet like a madman, I can eventually cause a crash, but not a quit like that :/.

**Comment by NiLuJe** [comment]
In your specific instance, given the log, @ ~02:41:50, the SleepCover Open did not register *at all*: we don't get an input event, and it's not a case of something failing to propagate it, because the kernel did not wakeup either.

The open/close pair later on (@ 02:42:02) *does* register properly, and in the right order.

So, at least for that specific log, there's nothing we can do about it, it's either not our bug, or a misaligned magnet ;).

**Comment by allati** [comment]
Thanks for all your effort trying to reproduce the issue! It can very well be hardware problem, it just does look strange that it somehow coincides with KOReader going to sleep (most of the time).

A couple of questions if I may. Though I have like zero knowledge of lua or KOReader architecture, may it be possible that once execution enters Kobo's Device:suspend, no keyboard events will be processed until devices enters RAM sleep, where kernel will be waiting for keyboard events to wake up? It can happen that I'm just opening sleepcover in 2 seconds sleep time in
```
util.sleep(2)
logger.info("Kobo suspend: waited for 2s because of reasons...")
```
? In this case there will be no one to handle that event: KOReader is already not reading events at this point (we need to leave Device:suspend to return to the main event loop and kernel is not yet reading events, so sleepcover open events are lost.

I try to hack code, but all I do is generating crashes :)

**Comment by NiLuJe** [comment]
Yep, in this specific instance, I'd be willing to chalk that one to the kernel's own suspend routine and unlucky timing ;).

**Comment by allati** [comment]
What if 2 seconds sleep is replaced with keyboard event reading for the same 2 seconds? This way suspend procedure can be aborted if sleepcover is opened. To better illustrate I've applied these changes here: allati/koreader@05c52e280305720b8f885a40e705b3cd1765498c. Not sure how much I've broken that way, but in this specific case KOReader now wakes normally.

**Comment by NiLuJe** [comment]
That 2s sleep is there to prevent deadlocks on a number of weird and not-so-weird situations, so i'm afraid it'll have to stay ;).

**Comment by allati** [comment]
Understood. Nothing can be done on this issue then, so better to close it. Thanks anyway!

**Comment by allati** [comment]
I've been using patched version for the last 3 week and I haven't encountered any deadlocks or increased battery usage (perceived battery usage at least). In addition to fixing case 3, cases 1 and 2 seems to be fixed as well: KOReader now always wakes on the first sleepcover open, even after long sleep. @NiLuJe, do you remember in which cases Kobo devices deadlocked without additions 2 seconds waits during sleep preparation? Had it been easy to reproduce? I know it was long ago, but still... If I can reproduce deadlock without 2 seconds additional sleep and if it goes away if instead of 2 seconds sleep we read key input for 2 seconds, then it would mean it is relatively safe use the latter algorithm.

And would it be possible to refactor Kobo device module to allow selecting which sleep algorithm to use? Via some "expert" settings in the UI or by editing configuration files directly. I've been trying to add such support through plugin to avoid patching KOReader constantly, but without success.

**Comment by NiLuJe** [comment]
It's very, very, very, very, very, very racy. So, basically, it could crash on basically anything depending on the alignment of the stars. Although anything < Mk. 7 is likely to be more susceptible to those kinds of issues.

(Mk. 7 devices are generally less finicky as far as suspend is concerned, and they're likelier not to do anything stupid, at least as far as "going to sleep" is concerned. It's one of the rare devices that I own that *will* suspend despite being in USBNet mode & charging, for instance).

Not quite sure about how easily that could be modularized, as that is a *very* finicky bit of black magic that we had a lot of trouble to get to behave on a large amount of different setups.

**Comment by NiLuJe** [comment]
Tl;DR:, while your patch certainly makes sense, and I can certainly see it working without any issue on a Forma, I'd be very wary to make it generic.

That said, this isn't in base, so it could be gated behind a setting. It might not be particularly pretty to achieve, code-wise, but it should be feasible ;).
You can try to piggyback on the "Ignore SleepCover events" options as an initial test, as I'll readily admit that groking how settings are handled is not necessarily straightforward as a first dip into KOreader ;).
